### DISCUSSION TITLE
How to add descriptions to search results

### DISCUSSION THREAD

**Discussion by jadhaidar** [junegunn/fzf#2982] [Q&A]
Hi there, I am trying to add descriptions to search results in the terminal similar to how it works when autocompleting git commands, for example:

<img width="643" alt="Screen Shot 2022-09-23 at 2 57 45 PM" src="https://user-images.githubusercontent.com/18439122/191946956-b66d8c98-1261-4a56-b8c2-ecf8b99dca29.png">

Specifically, I'd like add descriptions to some of the less frequently used git aliases that I have so I can quickly remember what they do without having to open the aliases file to see what they map to.

<img width="644" alt="Screen Shot 2022-09-23 at 2 59 04 PM" src="https://user-images.githubusercontent.com/18439122/191947309-17d182f9-91c4-443d-9671-c49c881f5f5e.png">

I did try to search for answers but couldn't find much (searching for "descriptions" in the Q&A category returns 0 results), and I couldn't find anything mentioning this in the docs. But seeing that it works for the git commands for example then I imagine it should be doable for other commands as well no?

**Comment by tg0h** [comment]
I keep all my aliases in multiple files in the $XDG_CONFIG_HOME/zsh/aliases directory

the function searches all my aliases and displays them on the left
- comments are shown in another colour
- the preview of the file is shown on the right
- I press enter if i need to edit the alias in nvim

dependencies needed
- rg (ripgrep) to search for aliases
- bat to preview the alias file and center the preview on the selected alias

<img width="1557" alt="Screenshot 2022-09-28 at 06 29 10 - kitty@2x" src="https://user-images.githubusercontent.com/7954426/192648902-31f53ad9-0647-48ec-9b41-06bb1fb339de.png">

## zsh function ae to display my aliases
```bash
#!bin/zsh
#filename of this is ae
export _fzf_home_alias_dir="$XDG_CONFIG_HOME/zsh/aliases"

_fzf_home_getAliases(){
  # match alias aliasName='blah'
  # match alias -g aliasName='blah'
  # match alias -g aliasName="blah"

  # \x1b is an escape code

  rg \
    --follow \
    --color always \
    --field-context-separator '' \
    --no-context-separator \
    -e "(^alias)(\s*)(-g)?(\s*)(?P<aliasName>(.*))(=)" -r '$aliasName ' \
    --field-match-separator ' ' \
    $_fzf_home_alias_dir \
    -g '!**/archive/*' \
    | gsed "s!$_fzf_home_alias_dir!!" \
    | gsed -E 's!(#)(\s*)!\1 !' \
    | gsed -E "s/#.*/"$'\x1b[38;2;82;96;255m'"&"$'\e[m'"/"
}

function ae(){
  # search home functions

  local result=$(
  _fzf_home_getAliases | fzf --ansi \
    --preview="fzf_home_alias_preview {1} {2}" \
    --bind "enter:execute(fzf_home_alias_edit {1} {2} < /dev/tty > /dev/tty 2>&1)+abort"
)
}

ae
```

## script fzf_home_alias_preview
this script provides the right hand preview of the file in the screenshot
```bash
#!/bin/zsh
# filename is fzf_home_alias_preview

# fzf_home_alias_preview <inputFile> <inputAlias>
# search for <inputAlias> in <inputFile>, find line number of inputAlias and provide to fzf-bat-preview
# needs to match
# alias test='test alias'
# alias test="test alias"
# alias -g test="test alias"
# alias -g test="test alias" # comments

local inputFile=$1
local inputAlias=$2

local aliasLineNumber
local file=$_fzf_home_alias_dir$inputFile

rg --only-matching --multiline --line-number --with-filename \
   -e "^alias(\s*)(-g)?(\s*)$inputAlias" $file \
   | rg '(?::)(\d+)(?::)' --only-matching --replace '$1' \
   | read -r aliasLineNumber _

fzf-bat-preview $_fzf_home_alias_dir$inputFile $aliasLineNumber
```
## script fzf-bat-preview to preview selected alias file with bat
centers the preview on the selected alias
```
#!/bin/zsh
# fzf-bat-preview <file> <highlightLineNumber>
# centres the preview with the specified win height

# number of lines to show
local WIN_HEIGHT=40

local file=$1
local highlightLine=$2
local winHeight=$WIN_HEIGHT

# subtract 3 for the header
local centreLine=$(( (winHeight - 3) / 2 ))

# if the highlight line is too small, start at 0 and not a negative number
local start=$(( highlightLine - centreLine ))
if [ $start -lt 0 ]; then
  start=0
fi

local end=$(( start + winHeight ))

bat --color always --highlight-line $highlightLine --line-range $start:$end --paging never "$file"
```

## script to edit with nvim if i press enter
```bash
#!/bin/zsh
# filename is fzf_home_alias_edit

  local inputFile=$1
  local inputAlias=$2
  cache=$XDG_STATE_HOME/dir-switcher/cache # store MRU files for each directory here

local file=$_fzf_home_alias_dir$inputFile

rg --only-matching --multiline --line-number --with-filename \
   -e "^alias(\s*)(-g)?(\s*)$inputAlias" $file \
   | rg '(?::)(\d+)(?::)' --only-matching --replace '$1' \
   | read -r aliasLineNumber _
 
# center the screen after opening the alias text file with nvim
  nvim $file +$aliasLineNumber -c 'normal! zz'
```

**Comment by oldfatcrab** [comment]
hey man I used `fzf-tab` to make it so...

not sure if that helps you:

[diff](https://github.com/Aloxaf/fzf-tab/compare/master...oldfatcrab:fzf-tab:master)

[demo video](https://youtu.be/L-Yml_kN9eU)
